nncf/quantization/algorithms/bias_correction/algorithm.py

```diff
@@ -227,40 +210,63 @@ def _get_subgraph_data_for_node(self, node: NNCFNode, nncf_graph: NNCFGraph) ->
         :param nncf_graph: NNCFGraph instance for graph analysis.
         :return: A dict with the list of the nodes for the subgraph input and statistics collection.
         """
-        stats_nodes, input_nodes, output_nodes = [], [], []
+        statistic_nodes, subgraph_input_nodes, subgraph_output_nodes = [], [], []
 
-        def traverse_to_layers_with_bias(node, output):
-            if node in output:
-                return True, output
-            if self._backend_entity.is_node_with_bias(node, nncf_graph):
-                output.append(node)
-                self._collected_stat_inputs.add(node.node_name)
-                activation_input = nncf_graph.get_input_edges(node)[0].from_node
-
-                output_nodes.append(activation_input)
-                return True, output
-            return False, output
-
-        def traverse_to_input_layers(node, output):
-            if node in output + input_nodes:
-                return True, output
-            if node.node_name in self._collected_stat_inputs and node not in stats_nodes:
-                output.append(node)
-                return True, output
-            return False, output
+        def fill_statistic_nodes(node):
+            # A small hack to speed up graph traversal.
+            if node in statistic_nodes or node in visited_nodes:
+                return
+            visited_nodes.append(node)
 
+            # If we found a node with bias, we have to collect it as a statistic node,
+            # and its input for _collected_stat_inputs_map,
+            # which will be used during the collection of statistics for the next node.
+            if self._backend_entity.is_node_with_bias(node, nncf_graph) and self._backend_entity.is_quantized_weights(
+                node, nncf_graph
+            ):
+                statistic_nodes.append(node)
+                activation_node, output_port_id = self._get_activation_node_and_port(node, nncf_graph)
+                subgraph_output_nodes.append(activation_node)
+                self._collected_stat_inputs_map[node.node_name] = (activation_node.node_name, output_port_id)
+                return
+
+            for next_node in nncf_graph.get_next_nodes(node):
+                fill_statistic_nodes(next_node)
+
+        def fill_subgraph_input_nodes(node):
+            # A small hack to speed up graph traversal.
+            if node in subgraph_input_nodes or node in visited_nodes:
+                return
+            visited_nodes.append(node)
+
+            # Since we need to find the inputs for the subgraph,
+            # we can take only those layers for which we have already collected statistics.
+            if node.node_name in self._collected_stat_inputs_map and node not in statistic_nodes:
+                subgraph_input_nodes.append(node)
+                return
+
+            for previous_node in nncf_graph.get_previous_nodes(node):
+                fill_subgraph_input_nodes(previous_node)
```

Comment:
There is really need to use recursive function? Can it be replaced to `for node in graph.get_all_nodes()`?